packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -661,12 +683,44 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       }
 
       const type = this.tsParseType();
+
+      const maybeLabeledElement =
+        type.type === "TSTypeReference" &&
+        !type.typeParameters &&
+        type.typeName.type === "Identifier";
+
+      const optional = this.eat(tt.question);
+
+      if (maybeLabeledElement && this.eat(tt.colon)) {
+        const labeledNode: N.TsTupleElementType = this.startNodeAtNode(type);
+        labeledNode.optional = optional;
+        labeledNode.label = type.typeName;
+        labeledNode.elementType = this.tsParseType();
+        return this.finishNode(labeledNode, "TSTupleElementType");
+      }
+
       // parses `TsType?`
-      if (this.eat(tt.question)) {
+      if (optional) {
+        // TODO(breaking): In this case, we should return a TSTupleElementType:
+        // {
+        //   type: "TSTupleElementType",
+        //   label: null,
+        //   optional: true,
+        //   elementType: ...
+        // }
```

Comment:
@JLHwung WDYT about using this AST?

Reply:
I tend to aligning naming convention with TSC, [example](https://ts-ast-viewer.com/#code/C4TwDgpgBAYg9nKBeKBtAzsATgSwHYDmANFAGY5aYBcUeArgLYBGEWJ6EAxnHgCYD8NTLkIkAdBKwRqUAIZ4QqALpKA3EA)

```js
{
  type: "NamedTupleMember";
  name: Identifier;
  optional: boolean;
  type: TSType;
}
```